Week 13 · Day 4 — Hugging Face Inference with GPT & BERT
Why this matters

Until now you trained on tiny toy datasets. But in practice, we use pretrained LLMs. Hugging Face gives you GPT-2, BERT, and hundreds of others ready for inference with just a few lines of code.

Theory Essentials

Hugging Face Transformers = hub for pretrained models.

Pipeline API makes inference easy (text-generation, fill-mask, sentiment).

GPT-2 → autoregressive generation (continue text).

BERT → fill-mask prediction (recover missing words).

Loading pretrained = avoids huge compute cost.

In [1]:
# Setup
from transformers import pipeline

# GPT-2: text generation
gen = pipeline("text-generation", model="gpt2")
print("GPT-2:\n", gen("The future of AI is", max_length=30, num_return_sequences=1)[0]["generated_text"])

# BERT: fill-mask
fill = pipeline("fill-mask", model="bert-base-uncased")
print("BERT:\n", fill("The capital of France is [MASK].")[0])


Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


GPT-2:
 The future of AI is a difficult one. We're not talking about a few new gadgets, or even a new computer, or something that we can easily replace with our own. We have to find ways to make sure that we deliver the things we want.

So, we've got to find ways to make sure that AI is able to deliver its tasks. We have to find ways to make sure that AI is able to provide us with real feedback when we're interacting with it, and that feedback is going to have an impact on which decisions we make, and which things we might implement, rather than just something that we're already doing.

And I think it's important that we have a system that's capable of delivering these kinds of things.

On how far we can go

I think if we're the first to get there, I think it would be very important to bring it directly to the real world.

We're talking about, for example, AI in the real world. We're talking about how we can create real-time feedback in real time.

I think that's what we're going to us

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


BERT:
 {'score': 0.4167883098125458, 'token': 3000, 'token_str': 'paris', 'sequence': 'the capital of france is paris.'}


1) Core (10–15 min)
Task: Run the pipelines and explain the difference in outputs.

GPT: Completes text. The future of AI is a difficult one. We're not talking about a few new gadgets, or even a new computer, or something that we can easily replace with our own. We have to find ways to make sure that we deliver the things we want.

So, we've got to find ways to make sure that AI is able to deliver its tasks. We have to find ways to make sure that AI is able to provide us with real feedback when we're interacting with it, and that feedback is going to have an impact on which decisions we make, and which things we might implement, rather than just something that we're already doing.

And I think it's important that we have a system that's capable of delivering these kinds of things.

On how far we can go

I think if we're the first to get there, I think it would be very important to bring it directly to the real world.

We're talking about, for example, AI in the real world. We're talking about how we can create real-time feedback in real time.


BERT: Fills in the blank. {'score': 0.4167883098125458, 'token': 3000, 'token_str': 'paris', 'sequence': 'the capital of france is paris.'}

2) Practice (10–15 min)
Task: Try your own prompts:

GPT: "Once upon a time,"

BERT: "The [MASK] is barking."

In [2]:
print("GPT-2:\n", gen("Once upon a time,", max_length=30, num_return_sequences=1)[0]["generated_text"])

print("BERT:\n", fill("The [MASK] is barking.")[0])

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


GPT-2:
 Once upon a time, when the world was so simple, and it was such a powerful thing that humanity was so strong, it was just the first time that a thing with such vast power of mind was able to be seen and heard.

"Oh, no, it's just a little bit, it's just the fact that the humans and monsters are so big and strong, and it's so that they can't even attack us, so it's just a little bit of luck. It's just luck."

The man who had finally taken this step had always been very stubborn and didn't show any mercy towards the beast, but he finally had the courage to admit that there was no chance of him and his followers to stop him.

"Do you know that this guy was able to stop that guy's attack? He made it his last chance to fight off the monster, so he was able to stop it. All of the other monsters are powerless to do anything, so if he's able to stop them, then it would be a big deal and he could make it to the bottom of the world."

The man who had finally taken this step said as he sl

3) Stretch (optional, 10–15 min)
Task: Change GPT parameters (temperature=1.2, top_k=50) and observe output diversity.
Higher temperature = more randomness.

In [3]:

print("GPT-2:\n", gen("AI is", max_length=30, temperature=1.2, top_k=50)[0]["generated_text"])


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


GPT-2:
 AI is working to expand its mobile analytics programme to more than 300 countries.

MOSCOW: So if something needs faster mobile growth, you're right in line with Google's goal of getting one billion smartphones globally by 2020. And Google told Bloomberg it hoped that in future all the Google services would be available in the U.S., including YouTube. How far will the company expect that to slow growth at this stage? Well we expect that to be even bigger by now, as we have more applications out on our mobile platforms than Google's on the desktop. If not, how soon - and what are the limits of its growing business - will it be possible to become the fastest and widest internet application network in the world?

PETER ELLISSON: We currently provide mobile users with the opportunity to access information, content, news and more, all without needing to access servers or hardware-accelerable server-side systems to be connected to your PCs. I think this has given us great flexibility

Output is more random

Mini-Challenge (≤40 min)

Build a dual-demo notebook:

User types input.

If input contains [MASK], call BERT pipeline.

Else, call GPT-2 pipeline.

Acceptance Criteria

Works for at least 5 examples.

Clear printed outputs.

Short explanation of why GPT vs BERT was chosen.

In [4]:
for i in range(5):

    text = input("Enter the input: ")

    if "[MASK]" in text:

        print("BERT:\n", fill(text)[0])
    
    else:
        print("GPT-2:\n", gen(text, max_length=30, num_return_sequences=1)[0]["generated_text"])


BERT:
 {'score': 0.9995494484901428, 'token': 2171, 'token_str': 'name', 'sequence': 'hello my name is charlie.'}


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


GPT-2:
 Is university worth it?

It is, however, worth it to graduate in an institution of higher learning, in a post-secondary education that is not geared toward the high-end of the humanities. If you're a student in a post-secondary institution and you are interested in learning and pursuing a degree at a higher level, the information that you'll find on the list is certainly worthwhile.

But it's not always worth it.


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


GPT-2:
 What is better working in your own proyect or going to uni?

I'm sure you've heard about the word "proyect". It's a term used for different things. It's not just about being able to do anything in a proyect, such as writing a chapter, or being able to write in a book. It's about having a good time and having fun. So I think I've got a lot of experience in it. But it's a good one.

There're different ways to write. Some have more to do with writing and writing short stories and some have more to do with writing and writing novels. It might not be as simple as you might think. But when you combine all those things, it's quite easy to write, and I think it's really fun!

Is it difficult to work in your own proyect?

I've read a lot of books about proyects and I'm pretty sure I've done a lot of it. It's easy to be a proyect, but it's not easy to be a proyect in the same way that I'd be if I didn't do the same things as I did.

So, what are some of the things that you can do with yo

Notes / Key Takeaways

Hugging Face = the easiest way to use pretrained LLMs.

GPT → text continuation; BERT → fill-mask.

Pipeline API abstracts tokenization + model inference.

Pretrained models save training cost and time.

Parameters (temperature, top-k) control generation style.

Reflection

Why is it important that pretrained models exist for GPT/BERT?

What risks come from using models “as-is” without fine-tuning?

1. Why is it important that pretrained models exist for GPT/BERT?
Because pretraining on massive text corpora gives models broad language knowledge (grammar, vocabulary, world facts). Fine-tuning then only needs small task-specific data, making advanced NLP possible for everyone without billions of dollars in compute.

2. What risks come from using models “as-is” without fine-tuning?
They may give generic or biased outputs, fail at your specific domain (e.g., medical, legal text), or produce misleading results. Without fine-tuning, the model doesn’t adapt to your context, so accuracy, reliability, and safety can be poor.